text/001x-battery_improvements.md

```diff
@@ -0,0 +1,183 @@
+	- [common: added voltage and current multipliers to BATTERY_STATUS #233](https://github.com/ArduPilot/mavlink/pull/233)
+    - [Add MAV_BATTERY_CHARGE_STATE_CHARGING #1068](https://github.com/mavlink/mavlink/pull/1068)
+
+  
+# Summary
+
+This RFC proposes:
+- a new `BATTERY_STATUS_V2` message that has a single cumulative voltage value rather than individual cell voltage arrays.
+- a new (optional) `BATTERY_VOLTAGES` message that can be scaled to as many cells as needed, and which reports them as faults.
+
+  
+# Motivation
+
+The motivation is to:
+- reduce the memory required for battery reporting ([anecdotal evidence](https://github.com/ArduPilot/mavlink/pull/233#issuecomment-976197179) indicates cases with 15% of bandwidth on a default RF channel).
+- Provide a cleaner message and design for both implementers and consumers.
+
+The [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) has two arrays that can be used to report individual cell voltages (up to 14), or a single cumulative voltage, and which cannot be truncated.
+The vast majority of consumers are only interested in the total battery voltage, and either sum the battery cells or get the cumulative voltage.
+By separating the cell voltage reporting into a separate message the new battery status can be much smaller, and the cell voltages need only be sent if the consumer is actually interested.
+
+# Detailed Design
+
+There are three parts to the design:
+1. A more efficient status message
+2. A scalable battery voltage message.
+3. Mechanisms that allow the new messages to seamlessly coexist with the old one along with eventual deprecation of the older message.
+
+## BATTERY_STATUS_V2
+
+The proposed message is:
+
+```xml
+    <message id="???" name="BATTERY_STATUS_V2">
+      <description>Battery dynamic information. This should be streamed (nominally at 1Hz). Static battery information is sent in SMART_BATTERY_INFO.</description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the battery. INT16_MAX for unknown temperature.</field>
+      <field type="uint32_t" name="voltage" units="mV" invalid="[UINT32_MAX]">Battery voltage (total).</field>
+      <field type="int16_t" name="current" units="cA" invalid="UINT16_MAX">Battery current (through all cells/loads). Positive if discharging, negative if charging. UINT16_MAX: field not provided.</field>
+      <field type="int32_t" name="current_consumed" units="mAh" invalid="-1">Consumed charge, -1: Current consumption estimate not provided.</field>
+      <field type="uint8_t" name="percent_remaining" units="%" invalid="UINT8_MAX">Remaining battery energy. Values: [0-100], UINT32_MAX: Remaining battery energy is not provided.</field>
+      <field type="uint32_t" name="fault_bitmask" display="bitmask" enum="MAV_BATTERY_FAULT">Fault/health/ready-to-use indications.</field>
```

Comment:
@dakejahl Below compares UAVCAN and MAVLink types as I believe they are now. 

MAVLink types are primarily designed around minimizing size of data sent. So we tend to use uints over floats where the required resolution can be provided in that way. Note that by "required" I mean available and useful resolution. We will output mA if that is what sensors provide and if that is useful for calculations needed by GCS. But if power modules only output cA and calculations we need to make don't care about mA resolution, then we wouldn't use a float to capture that information.

Why am I saying this? I'm hoping you'll look at the resolutions we get below and the values and say they are sufficient. If you think that, for example, we need more than .01A resolution and we might need more than 327.67 A then I might have to revisit the type. 

UAVCAN | MAVLink
--- | ---
NA | uint8_t id 
float16 temperature             # [C]         |  int16_t temperature [cdegC] (max 327.67)
float16 current                 # [Ampere]    | int16_t current [cA]  - (max 327.67 A)
float16 voltage                 # [Volt]      | uint32_t voltage [mV] - huge
uint32 current_consumed         # [mAh] 	# 0 if unknown | int32_t current_consumed mAh
uint8 percent_remaining         # [%] 		# 0 empty, 255 full | uint8_t percent_remaining [%] 0 to 100, 255 is not used.
uint32 status_flags | uint32_t fault_bitmask

The only specific thing I see above is that mavlink uses the 0-100 for the `percent_remaining` and 255 for "not used", while UAVCAN uses 0-255 for the percentage range.

I tend to like the way MAVLink does it unless a less than 1% resolution is actually useful. I am not sure it ever is.

Reply:
Matching types between uavcan/mavlink seems less important since we can always cast. I think it's more important to ensure the type is large enough to hold the max/min value we might see now and in the future. That being said I'm not sure which type is better `int16` or `float16`. An `int16` is at least easier to do the math wrt to precision/range...

But an `int16` is probably too small to be future proof for centi-amp, so for the uavcan definitions I would prefer to use 32bit types for current and current_consumed.

@hendjoshsr71 made a good point about deriving amp-hours remaining from percent remaining and full charge capacity, more resolution is better. But we could still reserve 255 for unused if we wanted. 
https://github.com/dronecan/DSDL/pull/7#discussion_r836881886

Reply:
@dakejahl I've now responded to that comment  "about deriving amp-hours".
In summary it seems to me that its a good point that there is an issue, but whether you use an int8 or uint8 it is likely to still be a problem. 

Read the thread, but what I think UAVCAN should do is provide the right information to allow the remaining charge to be determine as accurately/usefully as possible. I was thinking perhaps  `charge_remaining` in mA hours. 
You would probably drop the percentage, and force a consumer that demands percentage remaining to calculate it - presumably using `100*charge_remaining/(charge_remaining+charge_consumed)`

I would propose MAVLink send `charge_remaining`. We might end up sending `battery_percentage` too to simplify things for consumers - but I wouldn't expect UAVCAN to necessarily do so since you won't be talking to OSDs/FPVs.

My logic here is that this is very like a car. Most drivers just want to be beeped when they are running low on power, but we all still appreciate the fuel gauge, which is generally just treated as a percentage that we intuitively map to how long before we'll have to fill up.

However some cars also estimate the amount of driving time you still have based on your fuel and estimated current consumption. Further, on a drone, you want to know the estimated power available so that you can do mission feasibility. 

No one does that very well in a car or drone, but we want to make it possible (hence the field). Do you think we provide enough if we also provide `charge_remaining`? My assumption is that the GCS should estimate the current consumption by integrating the current over time (reasonable?)

In summary, the message must support the future use case, and I don't think it sounds like `percentage_remaining` does that.

Does this make sense?

Reply:
W.r.t.  maximum current 327.67 A seems "large". I've asked @auturgy for feedback on whether it seems future proof, and best approach if it is not. For example, we might have this as a float or a uint_32, with units of mA rather than cA.

I'll add this as a "Question"

Reply:
I agree I think including `charge_remaining` is a good idea. But it might also be a good idea to leave the % remaining too for the same reasons you pointed out about OSDs, as well as GCSs that don't request/get the SMART_BATTERY_INFO for whatever reason

Reply:
@dakejahl OK, I will add it to the proposal, and let's see what happens.

Reply:
@dakejahl FWIW it feels wasteful that we use int32 rather than uint32 for the memory as we're always wasting half that memory. Is 32767 mAh big enough/reasonable?

If not, tempted to make these uint32s just in case. Thoughts?

Reply:
32767 mAh is certainly not big enough.

Reply:
Thanks @hendjoshsr71 I will update. I think 4294967295 might be big enough :-)

@dakejahl I added `charge_remaining`. While thinking about it, it realize it is kind of essential. The theory behind the original battery designs is ancient - pretty much it is that you _always_ start with a full battery, you get the current consumed, and that gives you a fairly accurate battery remaining. The battery_percentage is in support of that.

But while it is always good to fly with full batteries every time I think it is silly to assume that everyone will. With this field if the battery knows it can always provide that information. If a battery sends this then we can know that the amount is not full to start with.

PS I work MAVLink/PX4 on Weds, Thurs. I drop in and out on other days if I have spare hours, but you can't rely on me. Apologies if I am slow to respond.